### PR TITLE
Option to make track covariance Pos-def in packedcandidate

### PR DESCRIPTION
**Pull request by vlimant**
#### PR description:

as reported in BPH https://indico.cern.ch/event/1155820/contributions/4853223/ the packing scheme of the track covariance in MINIAOD introduces the possibility for the covariance matrix to not be positive-definite, with implications to vertex refitting.
The  PR adds a new method to PackedCanidate responsible for returning a reference to a pseudoTrack with positive defined covariance matrix.

#### PR validation:

a simple analyser reading packed candidate was run retrieving the `pseudoPosDefTrack` to check that the determinant is getting positive.

### CODE REVIEW

**Comment by mtosi** [comment]
@vlimant first of all, thanks ! I'm afraid the code needs to be forced to follow the code-format

**Comment by mandrenguyen** [comment]
please test

**Comment by elusian** [comment]
Hello

a quick note that was also suggested in a similar PR (#34446, which added a method to obtain a low quality cov matrix in case none was stored and initially used a parameter in `pseudoTrack` for that): since the track and the matrix are embedded in the `PackedCandidate`, if we add a parameter to `pseudoTrack` whoever calls it first decides whether it is posdef or not. If the pseudotrack of a candidate was previously requested without the new parameter, calling `pseudoTrack` with the new parameter will _not_ give a forced positive definite matrix.

This means that if you require a positive definite matrix you must _always_ call `resetPseudoTrack`.
Technically that isn't even enough, because it's a race condition. Even if you reset the track, another thread can call `pseudoTrack` without the parameter before you have the chance to call it with the parameter, which means you can still get a non posdef matrix.

What we ended up doing in the LUT PR was adding the modification as a method and not inside the unpacking (in this case it would mean adding a `makeTheCovMatrixPosDef` method that modifies the matrix stored in the `PackedCandidate`) and then adding an `EDProducer` that reads the `PackedCandidate` collection and creates a copy with the modification applied.
This does not suffer from any of the race conditions. It also does not bother anyone that did not request the posdef matrix, since it will be in a new collection.

File: DataFormats/PatCandidates/src/PackedCandidate.cc [hunk 1]
```diff
@@ -105,6 +108,43 @@ void pat::PackedCandidate::unpackCovariance() const {
...
     unpackCovarianceElement(*m, packedCovariance_.dlambdadz, 1, 4);
     unpackCovarianceElement(*m, packedCovariance_.dphidxy, 2, 3);
+
+    if (forcePosDef) {
+      //      std::cout<<"unpacking enforcing positive-definite cov matrix"<<std::endl;
+      //calculate the determinant and verify positivity
+      double det_(0);
+      bool computed_ = m->Det(det_);
+      //      std::cout<<"during unpacking, the determinant is: "<<det_<<std::endl;
+      if (computed_ && det_ < 0) {
```
Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc:118]: in mkFit context we apply a full Sylvester's criterion; computing determinants for sub-matrices. The extra cost is not that much compared to computing the det for the full matrix
https://github.com/cms-sw/cmssw/blob/17333a928fd7b69adc42695086b04fd8a55564ce/RecoTracker/MkFit/plugins/MkFitOutputConverter.cc#L321-L338

File: DataFormats/PatCandidates/src/PackedCandidate.cc [hunk 2]
```diff
@@ -105,6 +108,43 @@ void pat::PackedCandidate::unpackCovariance() const {
     unpackCovarianceElement(*m, packedCovariance_.dxydz, 3, 4);
     unpackCovarianceElement(*m, packedCovariance_.dlambdadz, 1, 4);
     unpackCovarianceElement(*m, packedCovariance_.dphidxy, 2, 3);
+
+    if (forcePosDef) {
+      //      std::cout<<"unpacking enforcing positive-definite cov matrix"<<std::endl;
+      //calculate the determinant and verify positivity
+      double det_(0);
+      bool computed_ = m->Det(det_);
```
Comment (slava77) [inline DataFormats/PatCandidates/src/PackedCandidate.cc:116]: please follow CMSSW style (lower case camelCaps, no trailing underscore in locals); the comment is for the full file

Comment (vlimant) [reply DataFormats/PatCandidates/src/PackedCandidate.cc]: thanks for the reminder. "camelCaps" rule for local could be added to http://cms-sw.github.io/cms_coding_rules.html (unless I could not find it)

File: DataFormats/PatCandidates/src/PackedCandidate.cc (hunk 1 above)
Comment (vlimant) [reply DataFormats/PatCandidates/src/PackedCandidate.cc]: thanks for the pointer indeed.

**Comment by elusian** [comment]
@vlimant 
It's not a matter of limited scope of interest, in fact other plugins not using the new parameter is worse, as they just need to read the collection before or at the same time as the user of this code to create problems.

In general, if something is globally cached state (as the matrix and track members are) you cannot have more than one way to compute it, or you get the race when different readers try to fill the cache with different versions of the value.

My suggestion is to move the new code from an if branch inside `unpackCovariance` to a new (non-const) method that takes the already unpacked covariance matrix and modifies it, which has to be called explicitly on a copy of the `PackedCandidate`.
Since it's a copy (with a new cache slot) there is no issue with having different cached values.

If the user then needs a posdef matrix, they can copy the `PackedCandidate`, call the new method on the copy, and then all calls to `pseudoTrack` will give a track with a correct matrix.

**Comment by vlimant** [comment]
@elusian : what I meant is that BPH will have an analyser/ntupler producing what they need, and since they know about the special need they will enable the option, and no-one gets fooled. If the "fix" gets in mainstream unpacking of PackedCandidate, then we'll make it the default for all.

**Comment by slava77** [comment]
type tracking

**Comment by slava77** [comment]
I think that the content of packedCandidates should be unique per release during reading. The current implementation is job/implementation-dependent.

the acceptable solutions could be
- either on-the-fly return of a copy (of as little as just a pseudoTrack or its covariance) that can be used also used a new collection with pos-forced values
- or actually force-pos for everyone during unpacking (without options)

**Comment by vlimant** [comment]
I think this version will be acceptable

**Comment by vlimant** [comment]
please test

**Comment by vlimant** [comment]
please test

File: DataFormats/PatCandidates/src/PackedCandidate.cc [hunk 3]
```diff
@@ -161,6 +165,60 @@ float pat::PackedCandidate::dz(const Point &p) const {
...
+    TVectorD eigenValues(5);
+    eigenCov.EigenVectors(eigenValues);
+    double minEigenValue = eigenValues.Min();
+    double delta = 1e-6;
+    if (minEigenValue < 0) {
+      for (int i = 0; i < 5; i++)
+        m(i, i) += delta - minEigenValue;
+    }
+
+    //    bool notPosDef = (!m.Sub<AlgebraicSymMatrix22>(0, 0).Det(det) || det < 0) ||
```
Comment (clacaputo) [inline DataFormats/PatCandidates/src/PackedCandidate.cc:203]: Please remove this line plus the 3 after

**Review by clacaputo — COMMENTED** [review]
IIUC, this PR adds a new method to `PackedCanidate` responsible for returning a reference to a pseudoTrack with positive defined covariance matrix.
Could you please update the PR description? It is still saying:
> The PR introduce a forcePosDef option to the pseudoTrack() accessor, to "fix" the covariance matrix.

**Comment by vlimant** [comment]
@clacaputo thanks indeed. all done.

**Comment by vlimant** [comment]
please test

**Comment by vlimant** [comment]
+1

**Comment by vlimant** [comment]
all good @clacaputo

**Comment by clacaputo** [comment]
+reconstruction

- new method `pseudoPosDefTrack` add to `PackedCanidate`
- no reco changes

**Comment by slava77** [comment]
It would be nice to backport this to 10_6_X.

@vlimant would you be available to do this?
Thanks.

**Comment by mmusich** [comment]
> 10.6 and everything in between ?

10.6 would be needed because it's the main release for UL analysis. I think this feature is innocent enough to be added to all of that is in- between. What do @cms-sw/orp-l2 think?

**Comment by slava77** [comment]
I guess 12_4 for Run3 2022 analyses. I'm not sure also about 12_6: was there a plan to run re-miniAOD here?

**Comment by perrotta** [comment]
> > 10.6 and everything in between ?
> 
> 10.6 would be needed because it's the main release for UL analysis. I think this feature is innocent enough to be added to all of that is in- between. What do @cms-sw/orp-l2 think?

Maybe it will not be really used in all them, but I'd add it to the release that can be still used in analyses: besides 10_6 for UL, 12_4 for 2022 pp, 12_5 for 2022 HI. 12_6 for re-nano and MC studies.

**Comment by mmusich** [comment]
> Maybe it will not be really used in all them, but I'd add it to the release that can be still used in analyses: besides 10_6 for UL, 12_4 for 2022 pp, 12_5 for 2022 HI. 12_6 for re-nano and MC studies.

   * 10_6 for UL -> covered in https://github.com/cms-sw/cmssw/pull/40526 
   * 12_4 for 2022 pp -> done in https://github.com/cms-sw/cmssw/pull/40556
   * 12_5 for 2022 HI -> done in https://github.com/cms-sw/cmssw/pull/40555
   * 12_6 for re-nano and MC studies -> this is covered in this PR

**Pull request merged**